src/test/java/com/google/firebase/auth/FirebaseAuthIT.java

```diff
@@ -519,17 +510,112 @@ public void testImportUsersWithPassword() throws Exception {
+          .setHandleCodeInApp(false)
+          .build());
+      Map<String, String> linkParams = parseLinkParameters(link);
+      assertEquals(ACTION_LINK_CONTINUE_URL, linkParams.get("continueUrl"));
+      String email = resetPassword(user.email, "password", "newpassword",
+          linkParams.get("oobCode"));
+      assertEquals(user.email, email);
+      // Password reset also verifies the user's email
+      assertTrue(auth.getUser(user.uid).isEmailVerified());
+    } finally {
+      auth.deleteUser(user.uid);
+    }
+  }
+
+  @Test
+  public void testGenerateEmailVerificationResetLink() throws Exception {
+    RandomUser user = RandomUser.create();
+    auth.createUser(new CreateRequest()
+        .setUid(user.uid)
+        .setEmail(user.email)
+        .setEmailVerified(false)
+        .setPassword("password"));
+    try {
+      String link = auth.generateEmailVerificationLink(user.email, ActionCodeSettings.builder()
+          .setUrl(ACTION_LINK_CONTINUE_URL)
+          .setHandleCodeInApp(false)
+          .build());
+      Map<String, String> linkParams = parseLinkParameters(link);
+      assertEquals(ACTION_LINK_CONTINUE_URL, linkParams.get("continueUrl"));
+      // There doesn't seem to be a public API for verifying an email, so we cannot do a more
+      // thorough test here.
+      assertEquals("verifyEmail", linkParams.get("mode"));
+    } finally {
+      auth.deleteUser(user.uid);
+    }
+  }
+
+  @Test
+  public void testGenerateSignInWithEmailLink() throws Exception {
+    RandomUser user = RandomUser.create();
+    auth.createUser(new CreateRequest()
+        .setUid(user.uid)
+        .setEmail(user.email)
+        .setEmailVerified(false)
+        .setPassword("password"));
+    try {
+      String link = auth.generateSignInWithEmailLink(user.email, ActionCodeSettings.builder()
+          .setUrl(ACTION_LINK_CONTINUE_URL)
+          .setHandleCodeInApp(false)
+          .build());
+      Map<String, String> linkParams = parseLinkParameters(link);
+      assertEquals(ACTION_LINK_CONTINUE_URL, linkParams.get("continueUrl"));
+      String idToken = signInWithEmailLink(user.email, linkParams.get("oobCode"));
+      assertFalse(Strings.isNullOrEmpty(idToken));
+      assertTrue(auth.getUser(user.uid).isEmailVerified());
+    } finally {
+      auth.deleteUser(user.uid);
+    }
+  }
+
+  private Map<String, String> parseLinkParameters(String link) throws Exception {
+    Map<String, String> result = new HashMap<>();
+    int queryBegin = link.indexOf('?');
+    if (queryBegin != -1) {
+      String[] segments = link.substring(queryBegin + 1).split("&");
+      for (String segment : segments) {
+        int equalSign = segment.indexOf('=');
+        String key = segment.substring(0, equalSign);
+        String value = segment.substring(equalSign + 1);
+        result.put(key, URLDecoder.decode(value, "UTF-8"));
+      }
+    }
+    return result;
+  }
+
+  private String randomPhoneNumber() {
```

Comment:
I'd probably put this next to `RandomUser`, or maybe even as a static method on that class, just to make it easier to find.